### PR TITLE
[lib] Extend host structure to support auth keys for reconfiguration.

### PR DESCRIPTION
**Pull request by MayyaSunil**
See issue #83

### CODE REVIEW

**Comment by tomaszmrugalski** [comment]
I saw you edited scripts in src/share/database/* I suppose it's a good time to explain how we do schema updates. Since this change introduces a new column in all databases, this is backward
incompatible change. Therefore we need to bump up the major schema version and reset minor version.

Unfortunately, we can't ride on top of existing upgrade schemas as those are already released in 1.4 and there are many users who already upgraded to 6.0 schema in mysql. Therefore a new upgrade script (6.0 to 7.0) is needed. If we decide to add some other things, we'll be able to extend this
script. We'll be able to do that until 1.5.0-beta is released, when the majority of users will upgrade.

Also, a note about dhcpdb_create.{mysql|pgsql|cql}. This is essentially an original 1.0 schema with all upgrade scripts appended. Once you add an upgrade script, you should put its queries at the end. If CQL is causing troubles (it's a weird environment), please ignore it for now and create a separate issue for it. I can do it myself or ask one of the Kea engineers to help with this.

The code didn't compile for me on Mac OS 10.12.6 with the following error:

```
  CXX      libkea_dhcpsrv_la-lease.lo
host.cc:75:10: error: no type named 'random_device' in namespace 'std'
    std::random_device rd;
    ~~~~~^
1 error generated.
```

I've fixed that by adding missing #include <random> header.

There were other compilation errors for CQL:

```
In file included from cql_lease_mgr.cc:20:
../../../src/lib/dhcpsrv/cql_lease_mgr.h:295:30: warning: 'getLeases6' overrides a member function but is not marked 'override'
      [-Winconsistent-missing-override]
    virtual Lease6Collection getLeases6(const DUID& duid) const;
                             ^
../../../src/lib/dhcpsrv/lease_mgr.h:436:30: note: overridden virtual function is here
    virtual Lease6Collection getLeases6(const DUID& duid) const = 0; 
                             ^
cql_lease_mgr.cc:2128:14: error: out-of-line definition of 'getLease6' does not match any declaration in 'isc::dhcp::CqlLeaseMgr'
CqlLeaseMgr::getLease6(const DUID& duid) const {
             ^~~~~~~~~
cql_lease_mgr.cc:2130:57: error: use of undeclared identifier 'DHCPSRV_CQL_GET_DUID'
    LOG_DEBUG(dhcpsrv_logger, DHCPSRV_DBG_TRACE_DETAIL, DHCPSRV_CQL_GET_DUID)
                                                        ^
cql_lease_mgr.cc:2136:14: error: redefinition of 'duid_data' with a different type: 'CassBlob' (aka 'vector<unsigned char>') vs 'std::string' (aka
      'basic_string<char, char_traits<char>, allocator<char> >')
    CassBlob duid_data(duid.getDuid());
             ^
cql_lease_mgr.cc:2129:17: note: previous definition is here
    std::string duid_data = duid.toText();
                ^
cql_lease_mgr.cc:2145:12: error: no viable conversion from returned value of type 'Lease6Collection' (aka 'vector<shared_ptr<isc::dhcp::Lease6> >') to
      function return type 'Lease6Ptr' (aka 'shared_ptr<isc::dhcp::Lease6>')
    return (result);
           ^~~~~~~~
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:358:21: note: candidate constructor not viable: no known conversion from 'Lease6Collection'
      (aka 'vector<shared_ptr<isc::dhcp::Lease6> >') to 'boost::detail::sp_nullptr_t' (aka 'nullptr_t') for 1st argument
    BOOST_CONSTEXPR shared_ptr( boost::detail::sp_nullptr_t ) BOOST_SP_NOEXCEPT : px( 0 ), pn()
                    ^
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:422:5: note: candidate constructor not viable: no known conversion from 'Lease6Collection'
      (aka 'vector<shared_ptr<isc::dhcp::Lease6> >') to 'const boost::shared_ptr<isc::dhcp::Lease6> &' for 1st argument
    shared_ptr( shared_ptr const & r ) BOOST_SP_NOEXCEPT : px( r.px ), pn( r.pn )
    ^
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:634:5: note: candidate constructor not viable: no known conversion from 'Lease6Collection'
      (aka 'vector<shared_ptr<isc::dhcp::Lease6> >') to 'boost::shared_ptr<isc::dhcp::Lease6> &&' for 1st argument
    shared_ptr( shared_ptr && r ) BOOST_SP_NOEXCEPT : px( r.px ), pn()
    ^
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:450:5: note: candidate template ignored: could not match 'shared_ptr' against 'vector'
    shared_ptr( shared_ptr<Y> const & r, typename boost::detail::sp_enable_if_convertible<Y,T>::type = boost::detail::sp_empty() )
    ^
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:484:5: note: candidate template ignored: could not match 'auto_ptr' against 'vector'
    shared_ptr( std::auto_ptr<Y> && r ): px(r.get()), pn()
    ^
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:516:5: note: candidate template ignored: could not match 'unique_ptr' against 'vector'
    shared_ptr( std::unique_ptr< Y, D > && r ): px( r.get() ), pn()
    ^
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:532:5: note: candidate template ignored: could not match 'unique_ptr' against 'vector'
    shared_ptr( boost::movelib::unique_ptr< Y, D > r ): px( r.get() ), pn()
    ^
/usr/local/include/boost/smart_ptr/shared_ptr.hpp:643:5: note: candidate template ignored: could not match 'shared_ptr' against 'vector'
    shared_ptr( shared_ptr<Y> && r, typename boost::detail::sp_enable_if_convertible<Y,T>::type = boost::detail::sp_empty() )
    ^
cql_host_data_source.cc:948:13: error: use of undeclared identifier 'key_'
            key_ = NULL_RESERVED_KEY;
            ^
cql_host_data_source.cc:1250:13: error: no matching constructor for initialization of 'isc::dhcp::IPv6Resrv'
    return (IPv6Resrv(type, IOAddress(reserved_ipv6_prefix_address_),
            ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/lib/dhcpsrv/host.h:80:5: note: candidate constructor not viable: no known conversion from 'const cass_int32_t' (aka 'const int') to
      'const std::string' (aka 'const basic_string<char, char_traits<char>, allocator<char> >') for 3rd argument
    IPv6Resrv(const Type& type,
    ^
../../../src/lib/dhcpsrv/host.h:62:5: note: candidate constructor not viable: requires at most 3 arguments, but 4 were provided
    IPv6Resrv(const Type& type,
    ^
../../../src/lib/dhcpsrv/host.h:38:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 4 were provided
class IPv6Resrv {
      ^
../../../src/lib/dhcpsrv/host.h:38:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 4 were provided
1 warning and 4 errors generated.
```

I presume you never tested this with Cassandra. I can't blame you. Cassandra is a weird environment. Don't worry about it for now.

**Comment by fxdupont** [comment]
Note on the long term a crypto service (vs boost) must be used to get the random value.

File: src/lib/dhcpsrv/host.h [hunk 1]
```diff
@@ -82,6 +101,14 @@ class IPv6Resrv {
         return (type_);
     }
 
+    /// @brief Returns the keys.
+    ///
+    /// Keys are used for signing the Reconfigure Message.
+    /// The accessor should ensure 128 characters
+    std::string getKeys() const {
```
[the remark this answers is not in the document]
Comment (MayyaSunil) [reply src/lib/dhcpsrv/host.h]: Yes, agreed. I didnt quite understand "hex format". AFAIK key is just a 16 byte long characters. If the admin has configured less than 16 bytes we shall append with 0 and store it. Could you please elaborate a bit.

File: src/lib/dhcpsrv/host.cc [hunk 2]
```diff
@@ -43,18 +57,37 @@ IPv6Resrv::set(const Type& type, const asiolink::IOAddress& prefix,
...
 
+std::string
+IPv6Resrv::getRandomKeyString() 
+{
+    std::array <char, 128> randomString;
+    
+    std::random_device rd;
+    boost::random::mt19937 gen(rd());
+
+    std::for_each(randomString.begin(), randomString.end() - 1,
```
[the remark this answers is not in the document]
Comment (MayyaSunil) [reply src/lib/dhcpsrv/host.cc]: The branch contains two commits , please let me know if you want to squash them.

**Comment by MayyaSunil** [comment]
Have addressed all the review comments.

**Comment by fxdupont** [comment]
Note I discussed with Tomek about the PRNG usage in this code. Clearly in the long term a crypto grade PRNG will be required: there is already a ticket to add a PRNG interface to the cryptolink library. Now this should be postponed until we both know when it will be needed and the deadline should be not too far... So for now we are fine with a boost PRNG.

File: src/lib/dhcpsrv/host.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomaszmrugalski) [reply src/lib/dhcpsrv/host.cc]: No, we don't do squashing in Kea.

**Comment by tomaszmrugalski** [comment]
> The code didn't compile for me on Mac OS 10.12.6 with the following error:
> 
>   CXX      libkea_dhcpsrv_la-lease.lo
> host.cc:75:10: error: no type named 'random_device' in namespace 'std'
>     std::random_device rd;
>     ~~~~~^
> 1 error generated.
> 
> I've fixed that by adding missing #include header.

This hasn't been addressed. Please add #include <random> to host.cc

File: src/lib/dhcpsrv/lease_mgr.h [hunk 3]
```diff
@@ -430,6 +430,11 @@ class LeaseMgr {
     /// @return Lease collection (may be empty if no IPv6 lease found).
     virtual Lease6Collection getLeases6() const = 0;
 
+    /// @brief Returns collection of lease for matching DUID
+    ///
+    /// @return Lease collection (may be empty if no IPv6 lease found for the DUID).
+    virtual Lease6Collection getLeases6(const DUID& duid) const = 0;
```
[the remark this answers is not in the document]
Comment (MayyaSunil) [reply src/lib/dhcpsrv/lease_mgr.h]: This was the precisely reason I had proposed for adding new index earlier :P.  The proposal was to add new query to the db for fetching the leases.

File: src/lib/dhcpsrv/host.cc [hunk 4]
```diff
@@ -19,17 +21,69 @@ using namespace isc::asiolink;
...
+AuthKey::AuthKey(void) {
+    authKey_ = AuthKey::getRandomKeyString();
+}
+
+std::string
+AuthKey::getRandomKeyString() {
+    std::array <char, AuthKey::KEY_LEN> randomString;
+    
+    std::random_device rd;
+    boost::random::mt19937 gen(rd());
```
[the remark this answers is not in the document]
Comment (MayyaSunil) [reply src/lib/dhcpsrv/host.cc]: Apologies for missing out.

File: src/lib/dhcpsrv/lease_mgr.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (MayyaSunil) [reply src/lib/dhcpsrv/lease_mgr.h]: Such an implementation is available for cql , the optimised version will come out in a seperate branch soon.

**Comment by tomaszmrugalski** [comment]
Ok, I have re-reviewed your latest changes. There were still couple things to do: the getLeases6() should have been removed, there were errors in upgrade scripts (not sure whether they were side effect of the rebase or the problem on the branch itself) and couple lines were too long.

Anyway, pushed those changes to our internal master. Should pop up on github in an hour or so.

Thank you for writing this code.

**Comment by tomaszmrugalski** [comment]
Oh, forgot to mention that I compiled the code on Ubuntu 18.04 with all MySQL, PgSQL and CQL backends enabled. Tests took a long time, but they passed eventually.

**Pull request closed without merging**